File: java/Algorithms/src/algorithms/Merge.java
package algorithms;

import static algorithms.HelperMethods.*;

public class Merge {

    private static final int CUTOFF = 7;

    private static void merge(Comparable[] a, Comparable[] aux, int lo, int mid, int hi) {

        assert isSorted(a, lo, mid);
        assert isSorted(a, mid, hi);

        for (int k = lo; k <= hi; k++) {
            aux[k] = a[k];
        }

        int i = lo, j = mid + 1;
        for (int k = lo; k <= hi; k++) {
            if (i > mid) {
                a[k] = aux[j++];
            }
            else if (j > hi) {
                a[k] = aux[i++];
            }
            else if (less(aux[j], aux[i])) {
                a[k] = aux[j++];
            }
            else {
                a[k] = aux[i++];
            }
        }

        assert isSorted(a, lo, hi);

        printArr(a);
    }

    private static void sort(Comparable[] a, Comparable[] aux, int lo, int hi) {
        if (hi <= lo) {
            return;
        }

        /*
        if (hi <= lo + CUTOFF -1) {
            Insertion.sort(a, lo, hi);
            return;
        }
        */
        int mid = lo + (hi - lo) / 2;
        sort(a, aux, lo, mid);
        sort(a, aux, mid + 1, hi);

        if (less(a[mid + 1], a[mid])) {
            merge(a, aux, lo, mid, hi);
        }
    }

    public static void sort(Comparable[] a) {
        int N = a.length;
        Comparable[] aux = new Comparable[N];
        sort(a, aux, 0, N - 1);
    }

    public static void main(String[] args) {
        Integer[] a = generateRandomArray(100, 100);
        sort(a);
    }
}

File: dotnet/Algorithms/Algorithms/Merge.cs
using System;
using System.Collections;
using System.Diagnostics;

namespace Algorithms {

    public static class Merge<T> where T : IComparable {

        public static void Sort(T[] a) {
            var aux = new T[a.Length];
            for (int i = 0; i < a.Length; i++) {
                aux[i] = a[i];
            }
            Sort(a, aux, 0, a.Length - 1);
        }

        public static void Sort(T[] a, IComparer c) {
            var aux = new T[a.Length];
            Sort(a, aux, 0, a.Length - 1, c);
        }

        private static void Sort(T[] a, T[] aux, int lo, int hi) {
            if (hi <= lo) {
                return;
            }
            int mid = (lo + hi) / 2;
            Sort(a, aux, lo, mid);
            Sort(a, aux, mid + 1, hi);
            MergeArray(a, aux, lo, mid, hi);
        }

        private static void Sort(T[] a, T[] aux, int lo, int hi, IComparer c) {
            if (hi <= lo) {
                return;
            }
            int mid = (lo + hi) / 2;
            Sort(a, aux, lo, mid, c);
            Sort(a, aux, mid + 1, hi, c);
            MergeArray(a, aux, lo, mid, hi, c);
        }

        private static void MergeArray(T[] a, T[] aux, int lo, int mid, int hi) {
            Debug.Assert(IsSorted(a, lo, mid));
            Debug.Assert(IsSorted(a, mid + 1, hi));

            for (int k = lo; k <= hi; k++) {
                aux[k] = a[k]; 
            }

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++) {
                if (i > mid) {
                    a[k] = aux[j++];
                }
                else if (j > hi) {
                    a[k] = aux[i++];
                }
                else if (Less(aux[j], aux[i])) {
                    a[k] = aux[j++];
                }
                else {
                    a[k] = aux[i++];
                }
            }
            Debug.Assert(IsSorted(a, lo, hi));
        }

        private static void MergeArray(T[] a, T[] aux, int lo, int mid, int hi, IComparer c) {
            Debug.Assert(IsSorted(a, lo, mid, c));
            Debug.Assert(IsSorted(a, mid + 1, hi, c));

            for (var k = lo; k <= hi; k++) {
                aux[k] = a[k];
            }

            int i = lo, j = mid + 1;
            for (int k = lo; k <= hi; k++) {
                if (i > mid) {
                    a[k] = aux[j++];
                }
                else if (j > hi) {
                    a[k] = aux[i++];
                }
                else if (Less(aux[j], aux[i], c)) {
                    a[k] = aux[j++];
                }
                else {
                    a[k] = aux[i++];
                }
            }
            Debug.Assert(IsSorted(a, lo, hi, c));
        }

        public static void SortBottomUp(T[] a) {
            int length = a.Length;
            T[] aux = new T[length];
            for (int n = 1; n < length; n = n + n) {
                for (int i = 0; i < length - n; i += n + n) {
                    int lo = i;
                    int mid = i + n - 1;
                    int hi = Math.Min(i + n + n - 1, length - 1);
                    MergeArray(a, aux, lo, mid, hi);
                }
            }
            Debug.Assert(IsSorted(a));
        }

        private static bool IsSorted(T[] a) {
            return IsSorted(a, 0, a.Length - 1);
        }

        private static bool IsSorted(T[] a, int lo, int hi) {
            for (var i = lo + 1; i <= hi; i++) {
                if (Less(a[i], a[i - 1])) {
                    return false;
                }
            }
            return true;
        }

        private static bool IsSorted(T[] a, int lo, int hi, IComparer c) {
            for (var i = lo + 1; i <= hi; i++) {
                if (Less(a[i], a[i - 1], c)) {
                    return false;
                }
            }
            return true;
        }

        private static bool Less(IComparable v, IComparable w) {
            return v.CompareTo(w) < 0;
        }

        private static bool Less(object v, object w, IComparer c) {
            return c.Compare(v, w) < 0;
        }
    }
}